File: aural/pornhub-extractor.js
#!/usr/bin/env node

/**
 * Pornhub Data Extractor - Node.js Version
 *
 * This script extracts audio/video files and metadata from Pornhub using yt-dlp.
 * It downloads files, extracts metadata, and organizes data similar to other extractors.
 *
 * Usage: node pornhub-extractor.js <url>
 *
 * Requirements:
 * 1. Install yt-dlp: https://github.com/yt-dlp/yt-dlp
 */

const { spawn } = require("child_process");
const fs = require("fs").promises;
const path = require("path");

class PornhubExtractor {
  constructor(outputDir = "pornhub_data") {
    this.outputDir = path.resolve(outputDir);
  }

  async ensureOutputDir(uploaderDir) {
    try {
      await fs.access(uploaderDir);
    } catch (error) {
      await fs.mkdir(uploaderDir, { recursive: true });
      console.log(`📁 Created output directory: ${uploaderDir}`);
    }
  }

  async checkYtDlp() {
    return new Promise((resolve) => {
      const ytDlp = spawn("yt-dlp", ["--version"]);
      
      ytDlp.on("close", (code) => {
        if (code === 0) {
          console.log("✅ yt-dlp is available");
          resolve(true);
        } else {
          console.error("❌ yt-dlp not found. Please install yt-dlp:");
          console.error("   https://github.com/yt-dlp/yt-dlp");
          resolve(false);
        }
      });

      ytDlp.on("error", () => {
        console.error("❌ yt-dlp not found. Please install yt-dlp:");
        console.error("   https://github.com/yt-dlp/yt-dlp");
        resolve(false);
      });
    });
  }

  async extractFromUrl(url) {
    console.log(`🎯 Processing: ${url}`);
    
    // yt-dlp command with the exact format you specified
    const outputTemplate = path.join(
      this.outputDir,
      "%(uploader)s",
      "%(upload_date>%Y-%m-%d)s - %(id)s - %(fulltitle)s.%(ext)s"
    );

    const args = [
      "--output", outputTemplate,
      "--write-info-json",
      url
    ];

    console.log("🚀 Starting yt-dlp extraction...");
    console.log(`📂 Output directory: ${this.outputDir}`);

    return new Promise((resolve, reject) => {
      const ytDlp = spawn("yt-dlp", args);
      
      let stdout = "";
      let stderr = "";

      ytDlp.stdout.on("data", (data) => {
        const output = data.toString();
        stdout += output;
        // Show yt-dlp progress
        if (output.includes("[download]") || output.includes("[info]")) {
          process.stdout.write(output);
        }
      });

      ytDlp.stderr.on("data", (data) => {
        stderr += data.toString();
      });

      ytDlp.on("close", (code) => {
        if (code === 0) {
          console.log("✅ yt-dlp extraction completed successfully");
          resolve({ success: true, stdout: stdout });
        } else {
          console.error(`❌ yt-dlp failed with code ${code}`);
          console.error("STDERR:", stderr);
          reject(new Error(`yt-dlp failed: ${stderr}`));
        }
      });

      ytDlp.on("error", (error) => {
        console.error("❌ Failed to start yt-dlp:", error.message);
        reject(error);
      });
    });
  }

}

// CLI Interface
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.log("Usage: node pornhub-extractor.js <url>");
    console.log("Example: node pornhub-extractor.js https://www.pornhub.com/view_video.php?viewkey=123456");
    process.exit(1);
  }

  const url = args[0];
  const extractor = new PornhubExtractor();
  
  if (!(await extractor.checkYtDlp())) {
    process.exit(1);
  }
  
  try {
    const result = await extractor.extractFromUrl(url);
    console.log("\n✅ Extraction completed successfully");
  } catch (error) {
    console.error("\n❌ Extraction failed:", error.message);
    process.exit(1);
  }
}

if (require.main === module) {
  main().catch(error => {
    console.error("❌ Fatal error:", error.message);
    process.exit(1);
  });
}

module.exports = { PornhubExtractor };

File: aural/pornhub_extractor.py
#!/usr/bin/env python3
"""
Pornhub Data Extractor - Python Version

This script extracts audio/video files and metadata from Pornhub using yt-dlp.
It downloads files, extracts metadata, and organizes data similar to other extractors.

Usage: uv run python pornhub_extractor.py <url>
"""

import argparse
import sys
from pathlib import Path

import yt_dlp


class YtDlpLogger:
    """Custom logger for yt-dlp that uses emoji prefixes."""

    def debug(self, msg: str) -> None:
        # yt-dlp passes both debug and info through debug
        if msg.startswith("[debug] "):
            pass  # Skip debug messages
        else:
            self.info(msg)

    def info(self, msg: str) -> None:
        print(msg)

    def warning(self, msg: str) -> None:
        print(f"⚠️  {msg}")

    def error(self, msg: str) -> None:
        print(f"❌ {msg}")


class PornhubExtractor:
    def __init__(self, output_dir: str = "pornhub_data"):
        self.output_dir = Path(output_dir).resolve()

    def _progress_hook(self, d: dict) -> None:
        """Progress hook for yt-dlp downloads."""
        if d["status"] == "downloading":
            total = d.get("total_bytes") or d.get("total_bytes_estimate")
            downloaded = d.get("downloaded_bytes", 0)
            if total:
                percent = downloaded / total * 100
                speed = d.get("speed")
                speed_str = f"{speed / 1024 / 1024:.1f}MiB/s" if speed else "N/A"
                print(f"\r📥 Downloading: {percent:.1f}% at {speed_str}", end="")
        elif d["status"] == "finished":
            print("\n✅ Download complete, post-processing...")

    def extract_from_url(self, url: str) -> dict:
        """Extract content from Pornhub URL using yt-dlp."""
        print(f"🎯 Processing: {url}")
        print("🚀 Starting yt-dlp extraction...")
        print(f"📂 Output directory: {self.output_dir}")

        output_template = str(
            self.output_dir
            / "%(uploader)s"
            / "%(upload_date>%Y-%m-%d)s - %(id)s - %(fulltitle)s.%(ext)s"
        )

        ydl_opts = {
            "outtmpl": output_template,
            "writeinfojson": True,
            "logger": YtDlpLogger(),
            "progress_hooks": [self._progress_hook],
        }

        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            error_code = ydl.download([url])

        if error_code:
            raise RuntimeError(f"yt-dlp failed with error code {error_code}")

        print("✅ yt-dlp extraction completed successfully")
        return {"success": True}


def main() -> int:
    """CLI entry point."""
    parser = argparse.ArgumentParser(
        description="Extract video/audio from Pornhub using yt-dlp"
    )
    parser.add_argument("url", nargs="?", help="Pornhub URL to extract")
    parser.add_argument(
        "--output-dir",
        "-o",
        default="pornhub_data",
        help="Output directory (default: pornhub_data)",
    )

    args = parser.parse_args()

    if not args.url:
        print("Usage: uv run python pornhub_extractor.py <url>")
        print(
            "Example: uv run python pornhub_extractor.py "
            "https://www.pornhub.com/view_video.php?viewkey=123456"
        )
        return 1

    extractor = PornhubExtractor(output_dir=args.output_dir)

    try:
        extractor.extract_from_url(args.url)
        print("\n✅ Extraction completed successfully")
        return 0
    except Exception as error:
        print(f"\n❌ Extraction failed: {error}")
        return 1


if __name__ == "__main__":
    sys.exit(main())
